src/types.ts

```diff
@@ -1,7 +1,32 @@
 export type UnbindFn = () => void;
 
-export type Binding = {
-  type: string;
-  listener: EventListenerOrEventListenerObject;
+type AnyFunction = (...args: any[]) => any;
+
+type GetEventType<Target extends EventTarget, Type extends string> = Target extends unknown
+  ? `on${Type}` extends keyof Target
+    ? GetEventTypeFromListener<
+        // remove types that aren't assignable to `AnyFunction`
+        // so that we don't end up with union like `MouseEvent | Event`
+        Extract<Target[`on${Type}`], AnyFunction>
+      >
+    : Event
+  : never;
+
+type GetEventTypeFromListener<T extends AnyFunction> = T extends (this: any, event: infer U) => any
+  ? U extends Event
+    ? U
+    : Event
+  : Event;
+
+export type Binding<Target extends EventTarget = EventTarget, Type extends string = string> = {
+  type: Type;
+  listener: Listener<GetEventType<Target, Type>>;
   options?: boolean | AddEventListenerOptions;
 };
+
+export type Listener<Ev extends Event> =
```

Comment:
Can we tighten this up so that `this` is set correctly?

Reply:
Sure, we can.